plugins/flytekit-omegaconf/flytekitplugins/omegaconf/dictconfig_transformer.py

```diff
@@ -0,0 +1,194 @@
+from google.protobuf.json_format import MessageToDict, ParseDict
+from google.protobuf.struct_pb2 import Struct
+
+import omegaconf
+from flytekit import FlyteContext
+from flytekit.core.type_engine import TypeTransformerFailedError
+from flytekit.extend import TypeEngine, TypeTransformer
+from flytekit.models.literals import Literal, Scalar
+from flytekit.models.types import LiteralType, SimpleType
+from omegaconf import DictConfig, OmegaConf
+
+logger = logging.getLogger("flytekit")
+
+T = TypeVar("T")
+NoneType = type(None)
+
+
+def is_flatable(d: DictConfig) -> bool:
+    """Check if a DictConfig can be properly flattened and unflattened, i.e. keys do not contain dots."""
+    return all(
+        isinstance(k, str)  # keys are strings ...
+        and "." not in k  # ... and do not contain dots
+        and (
+            OmegaConf.is_missing(d, k)  # values are either MISSING ...
+            or not isinstance(d[k], DictConfig)  # ... not nested Dictionaries ...
+            or is_flatable(d[k])
+        )  # or flatable themselves
+        for k in d.keys()
+    )
+
+
+class DictConfigTransformer(TypeTransformer[DictConfig]):
+    def __init__(self, mode: typing.Optional[OmegaConfTransformerMode] = None):
+        """Construct DictConfigTransformer."""
+        super().__init__(name="OmegaConf DictConfig", t=DictConfig)
+        self.mode = mode
+
+    @property
+    def mode(self) -> OmegaConfTransformerMode:
+        """Serialisation mode for omegaconf objects defined in shared config."""
+        return SharedConfig.get_mode()
+
+    @mode.setter
+    def mode(self, new_mode: OmegaConfTransformerMode) -> None:
+        """Updates the central shared config with a new serialisation mode."""
+        SharedConfig.set_mode(new_mode)
+
+    def get_literal_type(self, t: Type[DictConfig]) -> LiteralType:
+        """
+        Provide type hint for Flytekit type system.
+
+        To support the multivariate typing of nodes in a DictConfig, we encode them as binaries (no introspection)
+        with multiple files.
+        """
+        return LiteralType(simple=SimpleType.STRUCT)
+
+    def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], expected: LiteralType) -> Literal:
+        """
+        Convert from given python type object ``DictConfig`` to the Literal representation.
+
+        Since the DictConfig type does not offer additional type hints for its nodes, typing information is stored
+        within the literal itself rather than the Flyte LiteralType.
+        """
+        # instead of raising TypeError, raising AssertError so that flytekit can catch it in
+        # https://github.com/flyteorg/flytekit/blob/60c982e4b065fdb3aba0b957e506f652a2674c00/flytekit/core/
+        # type_engine.py#L1222
+        assert isinstance(python_val, DictConfig), f"Invalid type {type(python_val)}, can only serialise DictConfigs"
+        assert is_flatable(
```

Comment:
Mabye it would be better to explicitly raise in case users use `python -O` which strips asserts.

Reply:
addressed in https://github.com/flyteorg/flytekit/pull/2299/commits/4fdb7e54e775cc16c5c2c510fd89a4b29135438f